### PR TITLE
Colors and seperate verts if vert color/normal differs

### PR DESCRIPTION
**Pull request by QuantumEntangledAndy**
This PR adds colors and changes how the voxel baker works.

Colors works by generating a texture (as a vec of RGBA u32) and updating the UVs so that the faces lie in the right place for this generated texture.

The changes to the voxel baker effect the hashing of position and also will insert a new vertex rather than reuse an old one unless position, color and normal are all the same.

End result is this:
Screenshot 2020-12-08 at 10 57 27

### CODE REVIEW

File: gaiku-3d/src/bakers/voxel/mod.rs [hunk 1]
```diff
@@ -210,27 +194,57 @@ impl Baker for VoxelBaker {
     }
 }
 
+/// Either get the vertex at this position or inser one.
```
Comment (norman784) [inline gaiku-3d/src/bakers/voxel/mod.rs]:
```suggestion
/// Either get the vertex at this position or insert one.
```

File: gaiku-3d/src/bakers/voxel/mod.rs [hunk 2]
```diff
@@ -210,27 +194,57 @@ impl Baker for VoxelBaker {
...
+/// Only returns an old vertex if the position normal and color are the same
+/// as the requsted one
```
Comment (norman784) [inline gaiku-3d/src/bakers/voxel/mod.rs]:
```suggestion
/// as the requested one
```

**Comment by norman784** [comment]
It looks promising, I will make time to test this and read properly the code.

File: gaiku-3d/src/bakers/voxel/mod.rs (hunk 1 above)
Comment (QuantumEntangledAndy) [reply gaiku-3d/src/bakers/voxel/mod.rs]: Yes I have a few typos there's even one in the title of the amethyst demo app... :(

File: gaiku-3d/src/bakers/voxel/mod.rs (hunk 2 above)
Comment (QuantumEntangledAndy) [reply gaiku-3d/src/bakers/voxel/mod.rs]: Anyways thanks for the spot :)

**Comment by norman784** [comment]
I cannot test this right now, when trying to run the terrain example the process stuck using 100% of the current thread.

**Comment by QuantumEntangledAndy** [comment]
the example takes a lot of cpu and time to get going. It is what I mean by needs optimisation :). I should probably swap to a faster paradigm but went for working first rather than fast.

**Comment by QuantumEntangledAndy** [comment]
Probably generating a 1024x1024 texture for each chunk is overkill when theres only a few colors in there. I shall put in some time checks and see where the bottle necks are though

**Comment by QuantumEntangledAndy** [comment]
So I just reduced the texture size to 32x32 and there no different in quality but it runs much faster :)

**Comment by QuantumEntangledAndy** [comment]
I've pushed some changes with the smaller tex sizes please have a go now :)

**Review by norman784 — APPROVED** [review]
Now it works, it's still slow but we can deal with that in another PR.

**Comment by QuantumEntangledAndy** [comment]
Some of that slowness might be amethyst. We should run it with amethysts slow checks disables and maybe with compiler optimisations and see the difference.

**Comment by QuantumEntangledAndy** [comment]
When running the example did you notice that it culls the objects regularly. I've been playing around with it and saw that it culled it when the bottom left corner (0,0,0) went out of view. Perhaps we could add some convenience functions that can move the center of the chunk to the actual center. Perhaps have centered as default?

**Comment by norman784** [comment]
> Some of that slowness might be amethyst. We should run it with amethysts slow checks disables and maybe with compiler optimisations and see the difference.

Yeah, I started working on a naive benchamark implementation #8, and saw that the baking times are low enought, also in the example all the gaiku stuff takes like 8 seconds (in my pc) only in the game state initialization.

> When running the example did you notice that it culls the objects regularly. I've been playing around with it and saw that it culled it when the bottom left corner (0,0,0) went out of view. Perhaps we could add some convenience functions that can move the center of the chunk to the actual center. Perhaps have centered as default?

We could try to change the position to the center, but I'd prefer to have only one of the implementation so we can maybe stick with the position been the center of the chunk, we could always add some convenience methods to get the coordinates of the boundaries or a specific corner.

**Comment by QuantumEntangledAndy** [comment]
For those reading this in the future you can get good speeds on the example with this

```bash
cargo run --example terrain --features metal,amethyst/no-slow-safety-checks --release
```

which will turn off amethysts slow checks and turn on compiler optimisations resulting in a much imporved runtime

**Pull request merged**